auth0/src/main/java/com/auth0/android/authentication/storage/CredentialsManager.java

```diff
@@ -140,6 +164,37 @@ public void onFailure(@NonNull AuthenticationException error) {
                 callback.onFailure(new CredentialsManagerException("An error occurred while trying to use the Refresh Token to renew the Credentials.", error));
             }
         });
+
+    }
+
+    private boolean hasScopeChanged(@NonNull String storedScope, @Nullable String requiredScope) {
+        if (requiredScope == null) {
+            return false;
+        }
+        String[] stored = storedScope.split(" ");
+        Arrays.sort(stored);
+        String[] required = requiredScope.split(" ");
+        Arrays.sort(required);
+        return stored != required;
+    }
+
+    private boolean willExpire(long expiresAt, long minTtl) {
+        long nextClock = getCurrentTimeInMillis() + minTtl * 1000;
+        return expiresAt <= nextClock;
+    }
+
+    private long calculateExpiresAt(@NonNull Credentials credentials) {
+        long expiresAt = credentials.getExpiresAt().getTime();
+
+        if (credentials.getIdToken() != null) {
+            JWT idToken = jwtDecoder.decode(credentials.getIdToken());
+            Date idTokenExpiresAtDate = idToken.getExpiresAt();
+
+            if (idTokenExpiresAtDate != null) {
+                expiresAt = Math.min(idTokenExpiresAtDate.getTime(), expiresAt);
+            }
+        }
+        return expiresAt;
     }
 
     /**
```

Comment:
`hasValidCredentials()` should have an overload that takes `minTtl`.

Reply:
See https://github.com/auth0/Auth0.swift/blob/master/Auth0/CredentialsManager.swift#L124

Reply:
@Widcket Thanks for pointing this! I've thought about this as well and got stuck thinking about the use case. When access/ID tokens are stored and have expired, the method would only return `true` if a refresh token is also stored, meaning that it can refresh it for new non-expired access/ID tokens. How will knowing the minTTL value affect the returned value if the refresh token is present to refresh it anyway?

Reply:
> How will knowing the minTTL value affect the returned value if the refresh token is present to refresh it anyway?

If there's not enough TTL left, then it should return `false` to inform that it does not satisfy the developer's TTL requirements.

>  the method would only return true if a refresh token is also stored

Why would a method that is meant to check the validity of the stored credentials concern itself with the presence/absence of the refresh token?

Reply:
Because the refresh token presence implies that the expired or not long-lived enough (minTTL) token can be renewed for a fresh pair. The end goal of the "hasCredentials" method is not to check if the _currently stored_ tokens are valid but if valid tokens would be obtainable when "retrieveCredentials" is called. The same is done in iOS [here](https://github.com/auth0/Auth0.swift/blob/master/Auth0/CredentialsManager.swift#L128).

Reply:
The refresh token presence means the credentials can be renewed, but not that they're currently valid. The `hasValid(minTTL:)` method doc comment states `Checks if a non-expired set of credentials are stored`, so the goal of that method is to check if the stored credentials are still valid or not indeed. Since `hasValid(minTTL:)` is a public method that can be used on its own, it should not do other things like check the refresh token presence/absence. Such a method should:
- Return `true` if the AT and ID Token have not expired, and if the AT will not expire within the TTL.
- Return `false` otherwise.

Reply:
If valid tokens are not obtainable because there is no refresh token, then the method that refreshes the credentials should fail, not `hasValid(minTTL:)`.

Reply:
That would be the outcome of applying the Principle of Least Astonishment, I think.

Reply:
We agreed to not change this method now. We can discuss for the next version.

Reply:
Actually, let's talk about this. Maybe now that we've split the methods it makes sense to add.

Reply:
Yes, it makes sense. Otherwise in one method (`getCredentials`) it's taking minTTL into account and in another (`hasValidCredentials`) it isn't. A customer that is using minTTL with `getCredentials` would reasonably expect `hasValidCredentials` to work the same way.

Reply:
Please add it.